libsql-server/src/namespace/meta_store.rs

```diff
@@ -0,0 +1,243 @@
+#![allow(dead_code)]
+
+use std::{collections::HashMap, path::Path, sync::Arc};
+
+use crate::{connection::config::DatabaseConfig, error::Error, Result};
+use parking_lot::Mutex;
+use rusqlite::Connection;
+use tokio::sync::{
+    mpsc,
+    watch::{self, Receiver, Sender},
+};
+
+use super::NamespaceName;
+
+type ChangeMsg = (NamespaceName, Arc<DatabaseConfig>);
+
+#[derive(Debug)]
+pub struct MetaStore {
+    changes_tx: mpsc::Sender<ChangeMsg>,
+    inner: Arc<Mutex<MetaStoreInner>>,
+}
+
+#[derive(Clone, Debug)]
+pub struct MetaStoreHandle {
+    namespace: NamespaceName,
+    inner: HandleState,
+}
+
+#[derive(Debug, Clone)]
+enum HandleState {
+    Internal(Arc<Mutex<Arc<DatabaseConfig>>>),
+    External(mpsc::Sender<ChangeMsg>, Receiver<Arc<DatabaseConfig>>),
+}
+
+#[derive(Debug)]
+struct MetaStoreInner {
+    // TODO(lucio): Use a concurrent hashmap so we don't block connection creation
+    // when we are updating the config. The config si already synced via the watch
+    // channel.
+    configs: HashMap<NamespaceName, Sender<Arc<DatabaseConfig>>>,
+    db: Connection,
+}
+
+/// Handles config change updates by inserting them into the database and in-memory
+/// cache of configs.
+async fn process(msg: ChangeMsg, inner: &Arc<Mutex<MetaStoreInner>>) -> Result<()> {
+    let (namespace, config) = msg;
+
+    let config_encoded = serde_json::to_vec(&config)?;
+
+    let inner = &mut inner.lock();
+
+    inner.db.execute(
+        "INSERT OR REPLACE INTO namespace_configs (namespace, config) VALUES (?1, ?2)",
+        rusqlite::params![namespace.to_string(), config_encoded],
+    )?;
+
+    let configs = &mut inner.configs;
+
+    if let Some(config_watch) = configs.get_mut(&namespace) {
+        config_watch.send_modify(|c| {
+            *c = config;
+        });
+    } else {
+        let (tx, _) = watch::channel(config);
+        configs.insert(namespace, tx);
+    }
+
+    Ok(())
+}
+
+#[tracing::instrument(skip(db))]
+fn restore(db: &Connection) -> Result<HashMap<NamespaceName, Sender<Arc<DatabaseConfig>>>> {
+    tracing::info!("restoring meta store");
+
+    db.execute(
+        "CREATE TABLE IF NOT EXISTS namespace_configs (
+            namespace TEXT NOT NULL PRIMARY KEY,
+            config BLOB NOT NULL
+        )
+        ",
+        (),
+    )?;
+
+    let mut stmt = db.prepare("SELECT namespace, config FROM namespace_configs")?;
+
+    let rows = stmt.query(())?.mapped(|r| {
+        let ns = r.get::<_, String>(0)?;
+        let config = r.get::<_, Vec<u8>>(1)?;
```

Comment:
I tried to change this to a borrowed version but since the row borrow is only for this closure we can't return it. I think since this is just a restore a string/vec<u8> is fine here, we can change if we find its really causing issues. Configs should be quite small anyways.